README.md
¿Qué vamos a hacer esta noche?
La ciencia suele usar ratones para crear los nuevos medicamentos del mañana. A pedido de una organización protectora de la vida ratónica, se nos pide elaborar un sistema de simulación de experimentos, para dejar tranquilos a Pinky y Cerebro.

De los ratones nos interesa modelar su nombre, su edad (en años), su peso, y las enfermedades que posee.

Por ejemplo:

--**Cerebro** es un ratón con 9 años, 0.2 kg de peso y tiene brucelosis, sarampión y tuberculosis.

--**Bicenterrata** es un ratón con 256 años, 0.2kg de peso, y completamente sano.

--**Huesudo** es un ratón de 4 años con 10kg de peso, y alta obesidad y sinusitis.

Modelar a los ratones mencionados.

Existen distintos tipos de hierbas que afectan (modifican) de diferentes maneras al ratón. Definir dichas hierbas:

--**hierbaBuena**, que rejuvenece al ratón a la raíz cuadrada de su edad.

  Por ejemplo, si a cerebro le doy hierbaBuena, se transforma en un ratón de 3 años.
  
--**hierbaVerde**, elimina las enfermedades que terminen de cierta forma.

  Por ejemplo, si a cerebro le doy la hierbaVerde del tipo “sis”, queda sólo con sarampión.
  
--**alcachofa**, hace que el ratón pierda peso en un 10% si pesa más de 2kg, sino pierde un 5%.

  Por ejemplo, un raton de 3 kg queda con 2,7 kg y cerebro queda con 0.19 kg. 
  
--**hierbaZort**, hace que el ratón se transforme en Pinky, perdiendo todas sus enfermedades y quedando con 0 años de edad.

--**hierbaDelDiablo**, hace que el ratón pierda 0.1kg (sin disminuir de 0) y elimina todas las enfermedades con menos de 10 letras. 



Medicamentos: Los medicamentos son la administración sucesiva de un conjunto de hierbas. Se pide crear los siguientes medicamentos para luego poder administrarlos en un ratón: 

--Hacer el **pondsAntiAge**, que es un medicamento que está hecho con 3 hierbas buenas y una alcachofa.
  Por ejemplo, si se lo administramos al ratón Bicenterrata, queda con 2 años y 0.19 kg 
  
--Hacer el **reduceFatFast**, (que viene en distintas potencias) y es un medicamento compuesto por una hierbaVerde de “obesidad” y tantas alcachofas como indique su potencia.
  Por ejemplo administrándole a Huesudo un reduceFatFast de potencia 1 hace que huesudo pase a pesar 9 kg y sólo quede con sinusitis. Si en lugar de la 1 le administramos un reduceFatFast de   potencia 2, pasa a pesar 8.1 kg y queda también solo con sinusitis.
  
--Hacer la **pdepCilina**, que es un medicamento que usa hierbasVerdes para curar todas las enfermedades infecciosas. Las enfermedades infecciosas son aquellas cuyo nombre termina de alguna de estas formas (utilizar esta constante):  sufijosInfecciosas = [ "sis", "itis", "emia", "cocos"]

Experimento: Los laboratorios antes de publicar un medicamento, lo prueban con distintos ratones para evaluar los resultados:

--Hacer la función que encuentra la cantidadIdeal. Recibe una condición y dice cuál es el primer número natural que la cumple.

--Saber si un medicamento lograEstabilizar una comunidad de ratones. Esto sucede cuando, luego de aplicarle el medicamento a todos los ratones de la comunidad, se elimina el sobrepeso y       todos tienen menos de 3 enfermedades. Un ratón tiene sobrepeso si pesa más de 1kg.

Diseñar el siguiente experimento: dado una comunidad de ratones, encontrar la potencia ideal del reduceFatFast necesaria para estabilizar la comunidad.

Queremos saber si un medicamento logra estabilizar una comunidad infinita. ¿Podemos saberlo? Responder en estos dos casos:

Si todos los ratones quedan con menos de 1kg y sin enfermedades. Justificar.

Si un ratón queda con 2kg y 4 enfermedades. Justificar.

Responder en base al ejercicio

--¿Qué cambios debería hacer para agregar una nueva hierba y construir un medicamento con ella? ¿Habría que modificar las funciones existentes? 
--¿Qué concepto está involucrado en la pregunta anterior? ¿Para qué sirve en este caso?
  Si se cambia el modelo del ratón, por ejemplo, ahora queremos que se registre el peso de un ratón siempre en libras. ¿Qué funciones habría que modificar?	


ChangeLog.md
# Changelog for ratones-test

## Unreleased changes


src/Lib.hs
module Lib where
import Text.Show.Functions
import PdePreludat


laVerdad = True

type Nombre = String
type Edad = Number
type Peso = Number
type Enfermedad = String

data Raton = UnRaton {
    nombre :: Nombre,
    edad :: Edad,
    peso :: Peso,
    enferemedades :: [Enfermedad]
}

cerebro :: Raton
cerebro = UnRaton {
    nombre = "Cerebro",
    edad = 9,
    peso = 0.2,
    enfermedades = [brucelosis, sarampion, tuberculosis]
}

bicenterrata :: UnRaton
bicenterrata = UnRaton {
    nombre = "Bicenterrata",
    edad = 256,
    peso = 0.2,
    enfermedades = []
}

huesudo :: UnRaton
huesudo = UnRaton {
    nombre = "Huesudo",
    edad = 4,
    peso = 10,
    enfermedades = [obesidad, sinusitis] 
}


type Hierba = (Raton -> Raton)

hierbaBuena :: Hierba
hierbaBuena raton = rejuvenecer (sqrt (edad raton)) raton

rejuvenecer :: Edad -> Raton -> Raton
rejuvenecer anios raton = raton {edad = anios}


hierbaVerde :: (String -> Hierba)
hierbaVerde sufijo raton = raton {enfermedades = filter (not . terminaCon sufijo) (enfermedades raton)}


terminaCon :: String -> String -> Bool
terminaCon sufijo enfermedad = sufijo == drop (length enfermedad - length terminacion) enfermedad


alcachofa :: Hierba
alcachofa raton = perderPeso (peso raton * pesoAPerder raton) raton

pesoAPerder :: Raton -> Number
pesoAPerder raton | (peso raton) > 2 = 0.1
                  | otherwise = 0.05

perderPeso :: Peso -> Raton -> Raton
perderPeso pesoReducir raton = raton {peso = max (peso raton - pesoReducir) 0}


cambiarNombre :: Nombre -> Raton -> Raton
cambiarNombre nuevoNombre raton = raton {nombre = nuevoNombre}

perderEnfermedades :: Raton -> Raton
perderEnfermedades raton = raton {enfermedades = []}

hierbaZort :: Hierba
hierbaZort raton = cambiarNombre "Pinky" . rejuvenecer 0 . perderEnfermedades

eliminarEnfermedades :: Number -> Raton -> Raton
eliminarEnfermedades letrasNombre raton = raton {enfermedades = filter ((<= letrasNombre).lenght) (enfermedades raton)}

hierbaDelDiablo :: Hierba
hierbaDelDiablo = perderPeso 0.1 . eliminarEnfermedades 10


type Medicamento = Raton -> Raton

componer :: [Hierba] -> Medicamento
componer [] = id
componer (x:xs) = x . componer xs

pondsAntiAge :: Medicamento
pondsAntiAge = componer [hierbaBuena, hierbaBuena, hierbaBuena, alcachofa]

listaDeHierbas :: Number -> [Hierba]
listaDeHierbas potencia = replicate potencia alcachofa ++ [hierbaVerde "obesidad"]

reduceFatFast :: Medicamento
reduceFatFast = componer . listaDeHierbas

sufijosInfecciosas :: [String]
sufijosInfecciosas = ["sis", "itis", "emia", "cocos"]

pdepCilina :: Medicamento
pdepCilina = componer (map (hierbaVerde sufijosInfecciosas))


cantidadIdeal :: (Num a, Enum a) => (a -> Bool) -> a
cantidadIdeal condicion = head (filter condicion [1..])


ningunoConSobrepeso :: [Raton] -> Bool
ningunConSobrepeso ratones = all ((< 1) . peso) ratones

menosDe3Enfermedades :: [Raton] -> Bool
menosDe3Enfermedades ratones = all ((< 3) . lengh . enfermedades) ratones

lograEstabilizar :: Medicamento -> [Raton] -> Bool
lograEstabilizar medicamento ratones = ningunoConSobrepeso (map medicamento ratones) &&  menosDe3Enfermedades (map medicamento ratones)




src/Library.hs
module Library where
import Text.Show.Functions
import PdePreludat


laVerdad = True

type Nombre = String
type Edad = Number
type Peso = Number
type Enfermedad = String

data Raton = UnRaton {
    nombre :: Nombre,
    edad :: Edad,
    peso :: Peso,
    enfermedades :: [Enfermedad]
}


brucelosis,sarampion,tuberculosis,obesidad,sinusitis :: Enfermedad
brucelosis = "brucelosis"
sarampion = "sarampion"
tuberculosis = "tuberculosis"
obesidad = "obesidad"
sinusitis = "sinusitis"

cerebro :: Raton
cerebro = UnRaton {
    nombre = "Cerebro",
    edad = 9,
    peso = 0.2,
    enfermedades = [brucelosis, sarampion, tuberculosis]
}

bicenterrata :: Raton
bicenterrata = UnRaton {
    nombre = "Bicenterrata",
    edad = 256,
    peso = 0.2,
    enfermedades = []
}

huesudo :: Raton
huesudo = UnRaton {
    nombre = "Huesudo",
    edad = 4,
    peso = 10,
    enfermedades = [obesidad, sinusitis] 
}


type Hierba = (Raton -> Raton)

hierbaBuena :: Hierba
hierbaBuena raton = rejuvenecer (sqrt (edad raton)) raton

rejuvenecer :: Edad -> Raton -> Raton
rejuvenecer anios raton = raton {edad = anios}


hierbaVerde :: (String -> Hierba)
hierbaVerde sufijo raton = raton {enfermedades = filter (not . terminaCon sufijo) (enfermedades raton)}


terminaCon :: String -> String -> Bool
terminaCon sufijo enfermedad = sufijo == drop (length enfermedad - length sufijo) enfermedad


alcachofa :: Hierba
alcachofa raton = perderPeso (peso raton * pesoAPerder raton) raton

pesoAPerder :: Raton -> Number
pesoAPerder raton | (peso raton) > 2 = 0.1
                  | otherwise = 0.05

perderPeso :: Peso -> Raton -> Raton
perderPeso pesoReducir raton = raton {peso = max (peso raton - pesoReducir) 0}


cambiarNombre :: Nombre -> Raton -> Raton
cambiarNombre nuevoNombre raton = raton {nombre = nuevoNombre}

perderEnfermedades :: Raton -> Raton
perderEnfermedades raton = raton {enfermedades = []}

hierbaZort :: Hierba
hierbaZort raton = (cambiarNombre "Pinky" . rejuvenecer 0 . perderEnfermedades) raton

eliminarEnfermedades :: Number -> Raton -> Raton
eliminarEnfermedades letrasNombre raton = raton {enfermedades = filter ((<= letrasNombre).length) (enfermedades raton)}

hierbaDelDiablo :: Hierba
hierbaDelDiablo = perderPeso 0.1 . eliminarEnfermedades 10


type Medicamento = Raton -> Raton

componer :: [Hierba] -> Medicamento
componer [] = id
componer (x:xs) = x . componer xs

pondsAntiAge :: Medicamento
pondsAntiAge = componer [hierbaBuena, hierbaBuena, hierbaBuena, alcachofa]

listaDeHierbas :: Number -> [Hierba]
listaDeHierbas potencia = replicate potencia alcachofa ++ [hierbaVerde "obesidad"]

reduceFatFast :: Number -> Medicamento
reduceFatFast = componer . listaDeHierbas

sufijosInfecciosas :: [String]
sufijosInfecciosas = ["sis", "itis", "emia", "cocos"]

pdepCilina :: Medicamento
pdepCilina = componer (map hierbaVerde (sufijosInfecciosas))


-- cantidadIdeal :: (Number, Enum a) => (a -> Bool) -> a
-- cantidadIdeal condicion = head (filter condicion [1..])


ningunConSobrepeso :: [Raton] -> Bool
ningunConSobrepeso ratones = all ((< 1) . peso) ratones

menosDe3Enfermedades :: [Raton] -> Bool
menosDe3Enfermedades ratones = all ((< 3) . length . enfermedades) ratones

lograEstabilizar :: Medicamento -> [Raton] -> Bool
lograEstabilizar medicamento ratones = ningunConSobrepeso (map medicamento ratones) &&  menosDe3Enfermedades (map medicamento ratones)


experimento :: [Raton] -> Number
experimento comunidad =

